diff --git a/lib/windowsPtyAgent.js b/lib/windowsPtyAgent.js
index a358ffb177357e177661033c1b092f9c9d0e5f5a..c56f72bfd9e323f34522287e1fe34d75b8ffa374 100644
--- a/lib/windowsPtyAgent.js
+++ b/lib/windowsPtyAgent.js
@@ -85,6 +85,10 @@ var WindowsPtyAgent = /** @class */ (function () {
             writable: true
         });
         this._inSocket.setEncoding('utf8');
+        this._outSocket.once('close', function () {
+            _this._inSocket.destroy();
+            _this._conoutSocketWorker.dispose();
+        });
         if (this._useConpty) {
             var connect = this._ptyNative.connect(this._pty, commandLine, cwd, env, this._useConptyDll, function (c) { return _this._$onProcessExit(c); });
             this._innerPid = connect.pid;
diff --git a/src/windowsPtyAgent.ts b/src/windowsPtyAgent.ts
index d7054449516f0c9a62af351c2caa17331206d530..c8275e7e2f6665bf9c8adef9a020dd83b49fd052 100644
--- a/src/windowsPtyAgent.ts
+++ b/src/windowsPtyAgent.ts
@@ -114,3 +114,7 @@ export class WindowsPtyAgent {
       writable: true
     });
     this._inSocket.setEncoding('utf8');
+    this._outSocket.once('close', () => {
+      this._inSocket.destroy();
+      this._conoutSocketWorker.dispose();
+    });
